core/sr-primitives/src/lib.rs

```diff
@@ -526,6 +531,29 @@ impl codec::Encode for ApplyError {
 	}
 }
 
+#[derive(Eq, PartialEq, Clone, Copy, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize))]
+/// Reason why a dispatch call failed
+pub struct DispatchError {
+	/// Module index, matching the metadata module index
+	pub module: i8, // use i8 instead of u8 because u8 is not supported by parity-codec
+	/// Module specific error value
+	pub error: i8,
+	/// Optional error message
+	#[codec(skip)]
+	pub message: Option<&'static str>,
+}
+
+#[derive(Eq, PartialEq, Clone, Copy, Encode, Decode)]
+ #[cfg_attr(feature = "std", derive(Debug, Serialize))]
+ /// Outcome of a valid extrinsic application. Capable of being sliced.
+ pub enum ApplyOutcome {
```

Comment:
Why not make `Result<(), DispatchError>` the return value of `apply_extrinsic`?

Reply:
This is based on the old code no I don't know why it was not a `Result`
But is it idiomatic to have nested Result?

Reply:
Why does it need to be nested?

Reply:
One version of my code the `ApplyResult` is an enum with three variants: Success, DispatchError and ApplyError.
Currently it is `Result<ApplyOutcome, ApplyError>` and Ok means transaction is included in the block, Err means transaction is not included in the block.
When transaction is included in the block, there are two outcomes: transaction is fully applied, or some error happened and not applied and the only changes are fee charge.
If I make `type ApplyOutcome = Result<(), DispatchError>` then `ApplyResult` will be a nested Result, which may be fine.

Reply:
I would have done the following:
```
enum ApplyError {
    ModuleError(ModuleError),
    // All the other apply errors
}
```

So you can easily see what happened and can just use `Result<(), ApplyError>`.

Reply:
But existing code are expecting `Err` means transaction won't be included, which is not the case for module error. It can be changed but just means a lot more code changes in other places and I want to keep this already big PR as small as possible.
This is not a regression (the structure is basically the old code) so can we make an issue and do it in later if you think this is something have to be addressed?

Reply:
Okay, I looked again into the code.

Could we then go the following way:
```
enum ApplyOutcome {
    Success,
    Fail(MethodError),
}
```

Reply:
That is same as the current one?
```
 pub enum ApplyOutcome {
 	/// Successful application (extrinsic reported no issue).
 	Success,
 	/// Failed application (extrinsic was probably a no-op other than fees).
 	Fail(DispatchError),
 }

pub enum ApplyError {
	/// Bad signature.
	BadSignature = 0,
	/// Nonce too low.
	Stale = 1,
	/// Nonce too high.
	Future = 2,
	/// Sending account had too low a balance.
	CantPay = 3,
	/// Block is full, no more extrinsics can be applied.
	FullBlock = 255,
}

pub type ApplyResult = Result<ApplyOutcome, ApplyError>;
```

Reply:
Ohh :sweat_smile: yeah! Sorry!